packages/rich-text/src/component/use-anchor.js

```diff
@@ -20,8 +20,22 @@ import { useState, useLayoutEffect } from '@wordpress/element';
 function getFormatElement( range, editableContentElement, tagName, className ) {
 	let element = range.startContainer;
 
-	// If the caret is right before the element, select the next element.
-	element = element.nextElementSibling || element;
+	// One issues with format boundaries is that the caret can be right before
+	// the element while we make it look to be inside the element. So if the
+	// startContainer is a text node and the caret is at the end of the text,
+	// we should use the next sibling and the deepest first child so we can
+	// select the correct format element. Do not use nextElementSibling, it must
+	// otherwise you may be matching an element further away.
```

Comment:
```suggestion
	// An issue with format boundaries is that the caret can be immediately before
	// the element whilst it can appear to be **inside** the element. Therefore if the
	// startContainer is a text node and the caret is at the end of the text,
	// we should use the next sibling and the deepest first child so we can
	// select the correct format element. This avoids potential bugs where 
	// an element (such as another pre-existing format) is incorrectly matched.
	// See: https://github.com/WordPress/gutenberg/issues/53641.
```

Nit: some wording tweaks.

Reply:
@ellatrix Are you happy with this suggested change?